### DISCUSSION TITLE
Ability to set HOME directory on windows (aka ~)

### DISCUSSION THREAD

**Discussion by Gei0r** [nushell/nushell#16659] [Q&A]
### Question

Is there a way to set `$nu.home-path` on windows? It's seems to default to `%USERPROFILE%` (??), but I'd like to use a different directory. Setting the `HOME` environment variable didn't help.

**Comment by fdncred** [comment]
Not at this time.

**Comment by fdncred** [comment]
This has been added https://www.nushell.sh/book/configuration.html#nushell-configuration-with-env-nu-and-config-nu

**Comment by NotTheDr01ds** [comment]
@fdncred Should be reopened, I think.  As you've come to realize in the Discord thread, there's still no way to do this.

It's not just Windows either as there's no way to do it on Linux currently.

Other shells (Bash, Fish, etc.) respect the `HOME` environment variable when checking the `~` shortcut, but Nushell does not.

**Comment by NotTheDr01ds** [comment]
Thinking through this further, I'm not sure what the exact behavior in Nushell should be.  At the least, `~` should probably respect the `$env.HOME`, but apart from that I'm not sure:

* In Bash and Fish, it seems that `$HOME` is read on *startup* and then used for all config and data files. Bash will write history to that directory, and Fish is even "kind" enough to create ~1500 completion files into `$HOME/.local/share/fish/generated_completions` (plus a few more).

  This kind of makes sense on those shells since they pre-dated the XDG config spec.  Since we already support `XDG_CONFIG_HOME` on Nushell, I don't think that it should look for config in `$HOME`.  It's safer and providers more flexibility to keep `HOME` and `XDG_CONFIG_HOME` distinct.

* Even in Nushell, the Starship prompt immediately starts to use the new `HOME` definition in storing its `.cache/starship`, and it does this on every prompt.

**Comment by fdncred** [comment]
I reopened it. Is there a XDG specification for redefining HOME? I don't think so, and didn't see anything but I could've missed it. https://specifications.freedesktop.org/basedir-spec/latest/

**Comment by NotTheDr01ds** [comment]
> Is there a XDG specification for redefining HOME?

Interestingly, yes, but I had to "read between the lines" before I found it.  

The primary focus of the XDG spec is on where *files* should be stored and located, so there's no discussion of how `~` is treated when it is set.

But for the purposes of config files, yes.  And Fish and Starship are pretty much following it, apparently.  For instance, from the spec:

> If `$XDG_CONFIG_HOME` is either not set or empty, a default equal to `$HOME/.config` should be used.

So technically, changing `$HOME` should have a similar effect on the shell as changing `$XDG_CONFIG_HOME`.

Let's say:

* `XDG_CONFIG_HOME = "/tmp/foo"`:

  Then `/tmp/foo` will be used for config files.  It doesn't matter what `$HOME` is set to.

* `HOME = "/tmp/foo"` and `XDG_CONFIG_HOME` is unset or empty:

  Then `/tmp/foo/.config` will be used for config files.

So I guess I was wrong about how to handle config files with `HOME`.  There's definitely a "right" way to do it according to the XDG spec.  Whether or not we choose that path ... ?

**Comment by NotTheDr01ds** [comment]
I guess, under XDG, if a user wanted to just redefine `HOME` for the purpose of `~`, then they'd do that in the shell startup, so that it took effect *after* the config file locations were defined.  So XDG still allows separation of those "features".

**Comment by fdncred** [comment]
Seems like changing $HOME would be a bad idea. Also, $HOME doesn't exist on Windows.

**Comment by NotTheDr01ds** [comment]
> Seems like changing $HOME would be a bad idea.

I'm not sure how?  It's a fairly standard feature in other shells.  And the XDG spec seems to anticipate it.  Not that I'm saying it's a high-priority feature by any means, just that I don't understand the negatives around it, since other shells handle it.

I think the main reason for changing `$HOME` is just to be able to `cd ~` to a different directory (and also use `~` to that directory in other expansions).  But using it to calculate the default `XDG_CONFIG_HOME` per the spec doesn't seems like it fits with our current design, just with a minor addition.

> Also, $HOME doesn't exist on Windows.

Neither does `XDG_CONFIG_HOME` (on Linux nor Windows), but we support it.  I guess the question is what is used to lookup `~` on Windows.  That's probably the biggest area of work, I would think, but is it a major change?

**Comment by fdncred** [comment]
I'm betting it would break many things on Windows. Not sure about other operating systems.

**Comment by ofek** [comment]
When does `~` expand exactly?

```
~\Desktop
❯ echo ~
~

~\Desktop
❯ cd ~

~
❯ echo $env.PWD
C:\Users\ofek
```

**Comment by NotTheDr01ds** [comment]
@ofek Sorry, missed this earlier.

`~` is expanded whenever it is in an argument that is treated as a [`path` shape](https://www.nushell.sh/lang-guide/chapters/types/other_types/path.html#additional-language-notes) or a [`glob type`](https://www.nushell.sh/lang-guide/chapters/types/basic_types/glob.html).

The `cd` command accepts a `path` (from `help cd`), so it is automatically expanded.  Likewise, `ls` accepts a glob (`help ls`), so `ls ~` is automatically expanded.

I haven't looked at the source to be sure, but the bareword `~` is probably special-cased.  When bareword string appears as the *only* item on the commandline, it appears to be treated as a potential path for "auto-`cd`" if:

* It starts with a `~`
* And/or it ends with a path-separator (e.g., `/`)

If it is not a valid directory, then it is treated as a command attempt.

```nushell
~
# => auto-cd to home
~/.config
# => auto-cd to $HOME/.config
~/not-a-dir-or-file
# => Error: Not a built-in nor external
```

In all other cases, I believe it will be interpreted as a bareword string:

```nushell
print ~
# => ~
echo ~
# => ~
```

**Comment by Lux-Solaris** [comment]
> > Seems like changing $HOME would be a bad idea.
> 
> I'm not sure how? It's a fairly standard feature in other shells. And the XDG spec seems to anticipate it. Not that I'm saying it's a high-priority feature by any means, just that I don't understand the negatives around it, since other shells handle it.
> 
> I think the main reason for changing `$HOME` is just to be able to `cd ~` to a different directory (and also use `~` to that directory in other expansions). But using it to calculate the default `XDG_CONFIG_HOME` per the spec doesn't seems like it fits with our current design, just with a minor addition.
> 
> > Also, $HOME doesn't exist on Windows.
> 
> Neither does `XDG_CONFIG_HOME` (on Linux nor Windows), but we support it. I guess the question is what is used to lookup `~` on Windows. That's probably the biggest area of work, I would think, but is it a major change?

Now Here's the thing: I've used Elvish on Windows, and Elvish is a shell that allows users to configure its "home directory", which not only matters with the `~` outcome, but also shifts a lot of configuration-files' directory, including the `.config` directory. To be specific, it makes everything that were started in the shell thought that home directory is the configured one, instead of the original default one, which is `C:/Users/<username>`.

**Comment by bminer** [comment]
Windows is a mess in this regard. `%USERPROFILE%` is sometimes mapped to a network drive in corporate settings, etc. I personally feel, `C:\Users\<username>` is the *right* answer 99% of the time, but because others have varying opinions here, I think making it somehow configurable isn't a bad compromise. I dunno...

**Comment by fdncred** [comment]
This is why we've tried to make it easy to customize your home. It's especially easy on Windows because you can easily set environment variables outside of nushell like XDG_CONFIG_HOME.
